Modules/permissions.py

```diff
@@ -115,15 +120,20 @@ def require_permission(permission: Permission,
     def real_decorator(func):
         log.debug("inside real_decorator")
         log.debug(f"Wrapping a command with permission {permission}")
-        # TODO implement require_permission wrapper.
 
         @wraps(func)
-        async def guarded(bot, trigger):
-            if trigger.identified and trigger.hostname in _by_vhost.keys() \
-                    and _by_vhost[trigger.hostname] >= permission:
-                return await func(bot, trigger)
+        async def guarded(bot, context: 'Context'):
+            log.debug(f"context = {context}")
+            granted_permissions = NONE
+            for key, value in _by_vhost.items():
+                if context.user.hostname.endswith(key) and context.user.identified is True:
+                    granted_permissions = value
```

Comment:
This method of iterating over `_by_vhost` is imperfect. Simply grabbing `_by_vhost[context.user.hostname]` if it exists worked fine, and should be more clear and be more performant.

Asking whether the user's hostname ends with the vhost is fragile in any event, as, for example, this would cause overlap between `techrat.fuelrats.com` and `rat.fuelrats.com`, which would lead to a sever bug of the techrat not being able to execute a command with a required permission level higher than `RAT`, since the rat level comes before the techrat level in `_by_vhost` and python dicts are basically ordered these days.

Reply:
``` granted_permissions = value if (not granted_permissions or granted_permissions < value) ``` should always give you the highest eligible permission level, assuming I didn't messed up that python

Reply:
it does not work by grabbing the hostname, as the hostname is more than just the host mask. 
Normal users' hostname looks something like `{name}.{ident}@{role}.fuelrats.com` and admins are `{name}@{role}.fuelrats.com`. 

Therefore, this is the simplest method was to itterate over the known hosts and checking if the hostname ends with a mask.

Reply:
Err, we're both half right. You're right, I forgot that what I assume to be a person's fuelrats.com username being part of their vhost. `_by_vhost[hostname]` won't work as I had hoped. However what you call `name`, generally called ident or username is not part of the hostname and (unless pydle's whois does some really strange things) is stored seperately in `user.username`.

Either way, there is a severe bug here which needs to be fixed. I see the simplest solution as not breaking when a match has been found or, slightly more performantly, reversing the `_by_vhost` dict (and maybe also making it a list as we don't need it to be a dict any longer and they do come with some overhead). This would ensure that the highest permission level which passed the `endswith(vhost)` test is always chosen. This is still somewhat fragile though, as if we ever come into the situation of a lower permission level ending with the vhost of a higher one, this would allow the lower-permissioned rat to attain the higher permission level.

An entirely different approach, which would likely be much more concise and clean, would be to include information of whether the vhost is preceded by stuff in each `Permission` object and, if so, cut out everything up until the first dot.
```python
found_permission = None
for permission in _permission_levels: # wouldn't need _by_vhost any longer
  if permission.vhost == (user.hostname.split(".", 1)[1] if permission.prefixed_by_username else user.hostname):
    found_permission = permission
    break
```
Or, functionally, which I would prefer:
```python
try:
  permission = next(filter(lambda permission: permission.vhost == (user.hostname.split(".", 1)[1] if permission.prefixed_by_username else user.hostname), _permission_levels))
except StopIteration:
  permission = NONE
```

Reply:
@MHajoha What will always remain consistent with our vhosts is that they *always* end with `{role}.fuelrats.com`.

would there be a concise way of checking for the amount of periods in the string, and based off of that either strip at the first period, or at the @ ?

If you could write it like that, you would be able to avoid writing a boolean to determine if the vhost has an ident in it.

Reply:
Yeah that'd work, great idea. Could use `hostname.rsplit(".", 3)[1]` to get at that part. I'm all for it. @theunkn0wn1, what do you think? Might be worth a new issue.

Reply:
`str.endswith()` allows us to compare the end of the string against another string. we do not need to do any fancy splits